thirdparty/patchmatchnet/models/patchmatch.py

```diff
@@ -0,0 +1,951 @@
+        # group wise correlation
+        self.G = G
+
+        self.stage = stage
+
+        self.dilation = propagation_out_range
+        self.propagate_neighbors = propagate_neighbors
+        self.evaluate_neighbors = evaluate_neighbors
+        self.evaluation = Evaluation(self.G, self.stage, self.evaluate_neighbors, self.patchmatch_iteration)
+        # adaptive propagation
+        if self.propagate_neighbors > 0:
+            # last iteration on stage 1 does not have propagation (photometric consistency filtering)
+            if not (self.stage == 1 and self.patchmatch_iteration == 1):
+                self.propa_conv = nn.Conv2d(
+                    in_channels=self.propa_num_feature,
+                    out_channels=2 * self.propagate_neighbors,
+                    kernel_size=3,
+                    stride=1,
+                    padding=self.dilation,
+                    dilation=self.dilation,
+                    bias=True,
+                )
+                nn.init.constant_(self.propa_conv.weight, 0.0)
+                if self.propa_conv.bias:
+                    nn.init.constant_(self.propa_conv.bias, 0.0)
+
+        # adaptive spatial cost aggregation (adaptive evaluation)
+        self.eval_conv = nn.Conv2d(
+            in_channels=self.propa_num_feature,
+            out_channels=2 * self.evaluate_neighbors,
+            kernel_size=3,
+            stride=1,
+            padding=self.dilation,
+            dilation=self.dilation,
+            bias=True,
+        )
+        nn.init.constant_(self.eval_conv.weight, 0.0)
+        if self.eval_conv.bias:
+            nn.init.constant_(self.eval_conv.bias, 0.0)
+        self.feature_weight_net = FeatureWeightNet(num_feature, self.evaluate_neighbors, self.G)
+
+    def get_propagation_grid(
+        self, batch: int, height: int, width: int, offset: torch.Tensor, device: torch.device, img: torch.Tensor = None
+    ) -> torch.Tensor:
+        """Compute the offset for adaptive propagation
+
+        Args:
+            batch: batch size
+            height: grid height
+            width: grid width
+            offset: grid offset
+            device: decide the tensor on which device
+
+        Returns:
+            generated grid: in the shape of [batch, len(original_offset)*H, W, 2]
+        """
+        if self.propagate_neighbors == 4:
+            original_offset = [[-self.dilation, 0], [0, -self.dilation], [0, self.dilation], [self.dilation, 0]]
+        elif self.propagate_neighbors == 8:
+            original_offset = [
+                [-self.dilation, -self.dilation],
+                [-self.dilation, 0],
+                [-self.dilation, self.dilation],
+                [0, -self.dilation],
+                [0, self.dilation],
+                [self.dilation, -self.dilation],
+                [self.dilation, 0],
+                [self.dilation, self.dilation],
+            ]
+        elif self.propagate_neighbors == 16:

```

Comment:
what do 8 and 16 represent?

Reply:
define how different `original_offset` would be according to different number of neighbors to be sampled (4, 8, or 16)